js/js.go

```diff
@@ -1,28 +1,47 @@
-//  | []int32, []int        | Int32Array            | []int                           |
-//  | []uint8               | Uint8Array            | []uint8                         |
-//  | []uint16              | Uint16Array           | []uint16                        |
-//  | []uint32, []uint      | Uint32Array           | []uint                          |
-//  | []float32             | Float32Array          | []float32                       |
-//  | []float64             | Float64Array          | []float64                       |
-//  | all other slices      | Array                 | []interface{}                   |
-//  | arrays                | see slice type        | see slice type                  |
-//  | functions             | Function              | func(...interface{}) *js.Object |
-//  | time.Time             | Date                  | time.Time                       |
-//  | -                     | instanceof Node       | *js.Object                      |
-//  | maps, structs         | instanceof Object     | map[string]interface{}          |
+// Internalization
 //
-// Additionally, for a struct containing a *js.Object field, only the content of the field will be passed to JavaScript and vice versa.
+// When values pass from Javascript to Go, a process known as internalization,
+// the following conversion table is applied:
+//
+//  |----------------+-------------------------+---------------+--------|
+//  | Go target type | Javascript source value | Translation   | Result |
+//  |----------------+-------------------------+---------------+--------|
+//  | string         | null                    | UTF16 -> UTF8 | ""     |
+//  |                | undefined               |               | ""     |
+//  |                | ""                      |               | ""     |
+//  |                | new String("")          |               | ""     |
+//  |                | "ok" †                  |               | "ok"   |
+//  |                | new String("ok") †      |               | "ok"   |
+//  |----------------+-------------------------+---------------+--------|
+//  | bool           | null                    | none          | false  |
+//  |                | undefined               |               | false  |
+//  |                | false †                 |               | false  |
+//  |                | new Boolean(false) †    |               | false  |
+//  |----------------+-------------------------+---------------+--------|
+//
+// Any source values not listed in this table cause a runtime panic for a given
+// target type if a conversion is attempted, e.g. a Javascript number value
+// being assigned to a string type Go variable.
+//
+// Source values annotated with † are generally applicable to all valid
+// values of the target type. e.g. for target type string, "ok" represents
+// all valid string values.
+//
+// Externalization
+//
+// When values pass from Go to Javascript, a process known as externalization,
```

Comment:
Thanks @flimzy - I should have mentioned this entire PR is still very much WIP.

In particular, I haven't even started on the `$externalize` side of things. Had some fairly interesting edge cases on the `$internalize` side of things